File: leetcode/python/799_champagne_tower.py
class Solution:
    def champagneTower(self, poured: int, query_row: int, query_glass: int) -> float:
        tower = [[0.0] * (query_row+2) for _ in range(query_row+2)]
        tower[0][0] = poured
        for row in range(query_row+1):
            for col in range(row+1):
                unit = (tower[row][col]-1)/2
                if unit > 0:
                    print(row+1, col)
                    tower[row+1][col] += unit
                    tower[row+1][col+1] += unit
        return min(tower[query_row][query_glass], 1.0)

s = Solution()
print(s.champagneTower(poured = 1, query_row = 1, query_glass = 1))
print(s.champagneTower(poured = 2, query_row = 1, query_glass = 1))
print(s.champagneTower(poured = 100000009, query_row = 33, query_glass = 17))

File: leetcode/rust/src/799_champagne_tower.rs
impl Solution {
    pub fn champagne_tower(poured: i32, query_row: i32, query_glass: i32) -> f64 {
        let mut query_row = query_row as usize;
        let mut tower = vec![vec![0.0; query_row+1]; query_row+1];
        tower[0][0] = poured as f64;

        for r in 0..query_row + 1 {
            for c in 0..r + 1 {
                let q = (tower[r][c] - 1.0) / 2.0;
                if q > 0.0 {
                    tower[r + 1][c] += q;
                    tower[r + 1][c + 1] += q;
                }
            }
        }

        return tower[query_row][query_glass as usize].min(1.0);
    }
}